### ISSUE TITLE
Checkstyle Unit Tests are coupled

### ISSUE THREAD

**Issue by HDouss** [teamed/qulice#547]
Checkstyle unit tests rely on availability of strings inside the logger output. Unit tests run on different threads (whether they are forked or not). It can happen that one test (i.e one thread) logs an error message that will be tested by another one and a false positive happens.
**Example :**
Test 1 checks and expects for error message "Checkstyle Error"

1. Test 1 (thread one) : arrives at point of execution L296 : after writer 1 is appended
2. Test 2 (thread two) : arrives at point of execution L312 : Checkstyle performs its checks and prints an output message "Checkstyle Error" on appenders of RootLogger (including writer 1)
3. Test 1 : continues execution, CheckStyle check fails due to another error and does not report "Checkstyle Error"

The Test 1 should fail but won't fail

**How to reproduce ?**
Add this test to `CheckstyleValidatorTest` : 
```java
    /**
     * CheckstyleValidator can report error.
     * @throws Exception In case of error
     */
    @Test
    public void reportsError() throws Exception {
        this.validateCheckstyle(
            "ReturnCount.java", false,
            Matchers.containsString(
                "Type Javadoc comment is missing an @param <T> tag."
            )
        );
    }
```
This test should fail because `ReturnCount.java` does not have parameters and Checkstyle should not produce the message `Type Javadoc comment is missing an @param <T> tag.` when checking `ReturnCount.java`.
The build could succeed (about 50% of the time) because `reportsErrorWhenParameterObjectIsNotDocumented` is causing the message to be appended on the `reportsError` (added test) writer.

**Comment by krzyk** [comment]
@davvd valid bug

**Comment by davvd** [comment]
> @davvd valid bug

@krzyk tag "bug" added

**Comment by davvd** [comment]
@hdouss many thanks, **15 mins** added to your acc for reporting this bug, pmt ID `73016327`

**Comment by krzyk** [comment]
@HDouss please extend the description to include also PMDValidatorTest, as it uses similar code and it can sometimes fail

**Comment by HDouss** [comment]
@krzyk You are right. I updated description.
However, the build should never fail because of the issue described. The described issue is about false positives, so it will never fail the build (but it rather should have failed it). If you have failing builds, it would come from other reasons as not flushing the writer.

**Comment by prahladyeri** [comment]
@HDouss @krzyk One workaround/resolution to this is the approach I used in PR #561. Make the test `classes` execute in parallel, not the individual `methods` by setting `parallel` to `classes` in surefire configuration. Classes are usually grouped according to their separate functionalities, so string outputs in different classes shouldn't clash with one another.

If that doesn't resolve the issue, however, make it single-threaded entirely:
    
    <parallel>classes</parallel>
    <forkCount>1</forkCount>
    <threadCount>1</threadCount>

Source: [StackOverflow](http://stackoverflow.com/questions/32987998/junit-and-surefire-parallel-tests-forkcount-threadcount/33047617)

**Comment by krzyk** [comment]
@prahladyeri but it is strange, as @HDouss wrote higher this can cause errors only in case when we assert that given message didn't appear, and one of the methods is producing it. And in your PR it looks like the message you are looking for didn't appear, which shouldn't happen.

**Comment by HDouss** [comment]
@prahladyeri I reproduced the issue on my machine, but I don't have enough informations/analysis to open a proper ticket.

**Comment by HDouss** [comment]
@krzyk We are talking about two different issues here. I suggest to let this issue focused on coupled tests that can induce false positives. The first one who could give enough informations about a scenario or an analysis on PMD failing tests should open a new issue.

**Comment by prahladyeri** [comment]
@krzyk 
> @prahladyeri but it is strange, as @HDouss wrote higher this can cause errors only in case when we assert that given message didn't appear

That's true, but there are some tests in `PMDValidator` class which depend on not finding a particular string in the log output (similar to @HDouss 's case). For example, in the method `doesNotComplainAboutConstantsInInnerClasses`:

    Matchers.not(Matchers.containsString("UnusedPrivateField")),

I believe this method was working properly until now when the total number of methods wasn't that high, but this new PR introduced 10 additional methods that ran in parallel, so these are now failing. This is explained by the fact that the build passes as soon as you make the cases run single-threaded.

@HDouss 
Sure. As I said, this is just one possible resolution, there could be multiple other ways of handling this.

**Comment by krzyk** [comment]
@davvd this is postponed

**Comment by davvd** [comment]
> @davvd this is postponed

@krzyk sure, thanks, I added "postponed" label to it

**Comment by davvd** [comment]
> @davvd this is postponed

@krzyk got it, someone else will be assigned soon

**Comment by prahladyeri** [comment]
@davvd @krzyk  Let me know whether I can try and fix this as per the approach I mentioned in my comment:

> One resolution to this is the approach I used in PR #561. Make the tests execute in parallel at the class level, not the individual methods level, by setting parallel to classes in surefire configuration.

**Comment by krzyk** [comment]
@prahladyeri I don't think this is a good approach. I would prefer to somehow get rid of the appender and introduce some other way of testing validator results.

**Comment by yegor256** [comment]
This issue is no longer reproducible on `master`. The architectural fix @krzyk suggested ("get rid of the appender and introduce some other way of testing validator results") has already landed: `CheckstyleValidator.validate()` and `PmdValidator.validate()` now return a per-call `Collection<Violation>` via instance-local listeners, and `CheckstyleValidatorTest` / `PmdValidatorTest` + `PmdAssert` assert directly against those return values. There is no shared `Appender` / `RootLogger` / `StringWriter` anywhere in `src/` any more, so parallel tests cannot pollute each other's output.

To guard against regressions I added a targeted test in commit d0d89e4a7 that encodes the exact scenario from the description — validating `ReturnCount.java` must not surface the `"Type Javadoc comment is missing @param <T> tag."` violation that belongs to `ParametrizedClass.java`. Under the original bug this would flake ~50% of the time; it now passes deterministically alongside all 53 checkstyle tests and 40 PMD tests.

Closing.
